### PR TITLE
Production pull

### PR DESCRIPTION
**Pull request by bubby932**
* Added Create Role endpoint.
* Added room image selection endpoint.
* Room IDs are now properly assigned.
* Added the setRoomPhoto permission.

### CODE REVIEW

File: administrator_dashboard.js [hunk 1]
```diff
@@ -409,24 +409,37 @@ async function updateRoom(room) {
...
+                "setRoomPhoto": true,
             }
         };
     } else {
         for(let i = 0; i < Object.keys(room.rolePermissions).length; i++) {
             const key = Object.keys(room.rolePermissions)[i];
             var value = room.rolePermissions[key];
 
-            if(typeof value.viewAndJoin != 'boolean') value.viewAndJoin = true;
+            if(typeof value.viewAndJoin != 'boolean') value.viewAndJoin = false;
```
Comment (bubby932) [inline administrator_dashboard.js:442]: rooms are PRIVATE by default not public

File: package.json [hunk 2]
```diff
@@ -21,18 +21,18 @@
     "express-rate-limit": "^6.7.0",
```
Comment (bubby932) [inline package.json:21]: file bumped by dependabot

File: routers/img.js [hunk 3]
```diff
@@ -106,6 +106,9 @@ router.post("/upload", uploadRateLimit, middleware.authenticateToken, async (req
         MetaData.infoPath = `/img/${MetaData._id}/info`;
         MetaData.filePath = `/img/${MetaData._id}`;
 
+        MetaData.takenInRoomId = room_id;
```
Comment (bubby932) [inline routers/img.js:109]: Fixed an issue where room ids were not assigned to photos.

File: routers/rooms.js [hunk 4]
```diff
@@ -128,7 +128,7 @@ router.route("/room/:room_id/subrooms/:subroom_id/versions/:version_id/download"
             const subroom = room.subrooms[subroom_id];
             
             if(version_id == 'latest') version_id = subroom.publicVersionId;
-            if(!subroom.versions[version_id].associated_file) return res.status(204).json({
+            if(!subroom.versions[version_id]?.associated_file) return res.status(204).json({
```
Comment (bubby932) [inline routers/rooms.js:131]: prevents a hard error

File: routers/rooms.js [hunk 5]
```diff
@@ -1159,6 +1159,218 @@ router.post('/new', authenticateTokenAndTag("Creative Tools Beta Program Member"
     }
 });
 
+router.put('/room/:id/roles/new', authenticateToken, canViewRoom, async (req, res) => {
```
Comment (bubby932) [inline routers/rooms.js:1162]: new role creation endpoint

File: routers/rooms.js [hunk 6]
```diff
@@ -1159,6 +1159,218 @@ router.post('/new', authenticateTokenAndTag("Creative Tools Beta Program Member"
...
+    } catch (ex) {
+        res.status(500).json({
+            code: "internal_error",
+            message: "An internal error occurred and we could not serve your request."
+        });
+        throw ex;
+    }
+});
+
+router.put("/room/:id/roles/:role_name/update", authenticateToken, requiresRoomPermission("managePermissions"), async (req, res) => {
```
Comment (bubby932) [inline routers/rooms.js:1230]: role permission update endpoint

File: routers/rooms.js [hunk 7]
```diff
@@ -1159,6 +1159,218 @@ router.post('/new', authenticateTokenAndTag("Creative Tools Beta Program Member"
...
+    } catch (ex) {
+        res.status(500).json({
+            code: "internal_error",
+            message: "An internal error occurred and we could not serve your request."
+        });
+        throw ex;
+    }
+});
+
+router.post("/room/:id/cover-image/set/:image_id", authenticateToken, requiresRoomPermission("setRoomPhoto"), async (req, res) => {
```
Comment (bubby932) [inline routers/rooms.js:1318]: room cover image set endpoint

**Review by bubby932 — COMMENTED** [review]
explained changes

File: .gitignore [hunk 8]
```diff
@@ -6,4 +6,7 @@ env.js
 nginx.conf
 data/audit.json
 admin.json
-keys/
\ No newline at end of file
+keys/
```
[the remark this answers is not in the document]
Comment (Bobrobot1) [reply .gitignore:9]: Lol

**Review by Bobrobot1 — APPROVED** [review]

**Pull request merged**